pycalphad/codegen/callables.py

```diff
@@ -93,67 +131,83 @@ def build_callables(dbf, comps, phases, model=None, parameters=None, callables=N
                                  .format(output, mod.__class__))
 
         if callables.get('callables', {}).get(name, False) and \
-                ((not build_gradients) or callables.get('grad_callables',{}).get(name, False)):
+                ((not build_gradients) or callables.get('grad_callables', {}).get(name, False)) and \
+                ((not build_hessians) or callables.get('hess_callables', {}).get(name, False)):
             _callables['callables'][name] = callables['callables'][name]
-            if build_gradients:
-                _callables['grad_callables'][name] = callables['grad_callables'][name]
-            else:
-                _callables['grad_callables'][name] = None
+            _callables['grad_callables'][name] = callables['grad_callables'].get(name, None)
+            _callables['hess_callables'][name] = callables['hess_callables'].get(name, None)
         else:
             # Build the callables of the output
             # Only force undefineds to zero if we're not overriding them
             undefs = {x for x in out.free_symbols if not isinstance(x, v.StateVariable)} - set(param_symbols)
             undef_vals = repeat(0., len(undefs))
             out = out.xreplace(dict(zip(undefs, undef_vals)))
-            build_output = build_functions(out, tuple([v.P, v.T] + site_fracs), parameters=param_symbols,
-                                           include_grad=build_gradients)
-            if build_gradients:
-                cf, gf = build_output
-            else:
-                cf = build_output
-                gf = None
+            build_output = build_functions(out, tuple(state_variables + site_fracs), parameters=param_symbols,
```

Comment:
`site_fracs` is sorted by construction in `Model.site_fractions`. The use of `variables` looks like it's obsolete now, probably from a cut-and-paste elsewhere. It's been removed and replaced by `site_fracs` for clarity, since the name "variables" is ambiguous and could refer to `state_variables + site_fracs` or just `site_fracs` alone.